src/app/CommandHandler.cpp

```diff
@@ -278,39 +301,41 @@ void CommandHandler::DecrementHoldOff()
 
     if (!mSentStatusResponse)
     {
-        if (!mExchangeCtx)
+        if (!mResponseSender.HasExchangeContext())
         {
             ChipLogProgress(DataManagement, "Skipping command response: exchange context is null");
         }
         else if (!IsGroupRequest())
         {
-            CHIP_ERROR err = SendCommandResponse();
+            CHIP_ERROR err = StartSendingCommandResponses();
             if (err != CHIP_NO_ERROR)
             {
                 ChipLogError(DataManagement, "Failed to send command response: %" CHIP_ERROR_FORMAT, err.Format());
+                mResponseSender.SendStatusResponse(Status::Failure);
+                Close();
+                return;
             }
         }
     }
 
+    if (mResponseSender.AwaitingStatusResponse())
+    {
+        // If we are awaiting a status response, we want to call Close() only once the response sender is done.
+        // Therefore, register to be notified when CommandResponseSender is done.
+        mResponseSender.RegisterOnResponseSenderDoneCallback(&mResponseSenderDone);
+        return;
+    }
     Close();
 }
 
-CHIP_ERROR CommandHandler::SendCommandResponse()
+CHIP_ERROR CommandHandler::StartSendingCommandResponses()
 {
-    System::PacketBufferHandle commandPacket;
-
     VerifyOrReturnError(mPendingWork == 0, CHIP_ERROR_INCORRECT_STATE);
     VerifyOrReturnError(mState == State::AddedCommand, CHIP_ERROR_INCORRECT_STATE);
-    VerifyOrReturnError(mExchangeCtx, CHIP_ERROR_INCORRECT_STATE);
-
-    ReturnErrorOnFailure(Finalize(commandPacket));
-    ReturnErrorOnFailure(
-        mExchangeCtx->SendMessage(Protocols::InteractionModel::MsgType::InvokeCommandResponse, std::move(commandPacket)));
-    // The ExchangeContext is automatically freed here, and it makes mpExchangeCtx be temporarily dangling, but in
-    // all cases, we are going to call Close immediately after this function, which nulls out mpExchangeCtx.
-
-    MoveToState(State::CommandSent);
+    VerifyOrReturnError(mResponseSender.HasExchangeContext(), CHIP_ERROR_INCORRECT_STATE);
 
+    ReturnErrorOnFailure(FinalizeInvokeResponseMessage(/* aHasMoreChunks = */ false));
```

Comment:
Suggest having 2 methods: `FinalizeLastInvokeResponseMessage` and `FinalizeIntermediateInvokeResponseMessage` to avoid the `/* aHasMoreChunks = */ ` everywhere.